Golden retriever refactor and UppyFile type improvements

mifi opened the pull request:
Probably best reviewed commit by commit.

I also split UppyFile into two intefaces distinguished by the `isRemote` boolean:
- LocalUppyFile
- RemoteUppyFile

Also:
- Removed the TagFile type
- Don't re-upload completed files - fixes #5930
- Clean up stored files on `complete` event *only* if *all* files succeeded (no failed files). this allows the user to retry failed files if the browser & upload get interrupted - fixes #5927, closes #5955
- Only set `isGhost` for non-successful files. it doesn't make sense for successfully uploaded files to be ghosted because they're already done. #5930

qxprakash commented on packages/@uppy/tus/src/index.ts:
```diff
@@ -539,7 +545,7 @@ export default class Tus<M extends Meta, B extends Body> extends BasePlugin<
 
     return {
-      ...file.remote?.body,
+      ...('remote' in file && file.remote.body),
```
This looks good , AI seems to be hallucinating here lol.

qxprakash commented on packages/@uppy/golden-retriever/src/index.ts:
```diff
@@ -56,386 +53,367 @@ export default class GoldenRetriever<
+
+      this.#patchMetadata({ files: filesWithoutBlobs })
+    }
+    // todo handle also blob changes here. currently we don't handle all blob changes, for example compressed image
```
https://github.com/transloadit/uppy/issues/6013

qxprakash commented:
Yeah sure 👍 , I didn't push any new code, just type changes , throttle.

mifi commented:
have now pushed a few new commits that addresses the issues and fixes a few TODOs in the code as well. I decided to ditch all the event handlers inside golden retriever and instead use the main state update event for all stored data, so we can be 100% sure that what we store is consistent with the actual state. i've also added back throttle, but it has to be done inside MetaDataStore or else we risk losing data.

qxprakash commented on packages/@uppy/golden-retriever/src/index.ts:
```diff
@@ -56,386 +54,362 @@ export default class GoldenRetriever<
+    this.#metaDataStore = new MetaDataStore({
       expires: this.opts.expires,
       storeName: uppy.getID(),
+      throttleTime: this.opts.throttleTime,
```
This changes the previous default throttling behavior of 500 ms. Now, users have the option to set it themselves. Should this be considered a breaking change?

mifi replied:
I didn't change the time of 500ms. the default is still 500ms. not sure why adding a new option would be a breaking change

qxprakash replied:
>I didn't change the time of 500ms. the default is still 500ms. not sure why adding a new option would be a breaking change

I missed looking into the constructor of `MetadataStore` ,  anyways it doesn't alter the default behaviour so not a breaking change, but we're still amending the public API no ?  

regarding whether we should give this option to the users or not: 

- Is the average user aware of the peculiar edge case you mentioned above? I’m not sure about that. I understood it because I spent time going through the code now. I’m not sure if an average user would do the same correct me if I’m wrong here.
- I’m fine either way, it’s just that if we decide to make this change, we’ll need to communicate the significance of this option clearly and also inform users about the above edge case.

qxprakash reviewed (approved):
I tested the changes locally they look good , you consolidated multiple operations into `#handleStateUpdate` it's really dense. secondly do we need to add new tests covering the nullable file data changes ?

mifi commented on private/dev/vite.config.js:
```diff
@@ -12,6 +12,20 @@ const config = {
+      input: {
+        main: './index.html',
+        sw: './sw.js',
+      },
```
https://github.com/transloadit/uppy.io/pull/386

mifi commented:
I've now updated the changesets. added an uppy.io PR here:  https://github.com/transloadit/uppy.io/pull/386

> you consolidated multiple operations into `#handleStateUpdate` it's really dense

you mean split it into multiple methods? `#handleStateUpdate` is 79 lines, I think it's not too bad

> secondly do we need to add new tests covering the nullable file data changes

which edge case are you thinking we should test? there are also more edge cases we should implement tests for here: https://github.com/transloadit/uppy/pull/5978/files#diff-57baae42738ccf74acb4ca6279ea3455582ad081a14b51b14bae146960c9ef8cR314-R317 but maybe we  should do it in a new PR?

Murderlon reviewed (approved):
Didn't test this locally but you spend a lot of time on it and Prakash did a deep dive review + local testing so I trust this is good now.

This should be resolved though before merging: https://github.com/transloadit/uppy/pull/5978#discussion_r2438792241

qxprakash commented:
> > you consolidated multiple operations into `#handleStateUpdate` it's really dense
> 
> you mean split it into multiple methods? `#handleStateUpdate` is 79 lines, I think it's not too bad

Yes it's fine , and let's talk about tests separately , this PR is getting long 😅
